### PR TITLE
Display Route

### PR DESCRIPTION
**Pull request by thienlnam**
### Fixed Issues

$ https://github.com/Expensify/App/issues/22708
PROPOSAL: 

### Tests

For engineers, if you're building and testing this PR in the development environment, you first need to run npm run configure-mapbox to setup the development environment.

1. Go to 'Request Money', 'Distance'
2. Add in a couple addresses
3. Verify that you see the route
4. Add stops
5. Verify that new routes are generated correctly
6. Remove one of the stops by editing the waypoint, and deleting all the text and pressing 'Enter'
7. Verify the new route is generated successfully

### Offline tests

Offline is known to not work at this time. You can save waypoints while offline but when coming online there will be an error. This is known at the time.

### QA Steps

1. Go to 'Request Money', 'Distance'
2. Add in a couple addresses
3. Verify that you see the route (You might have to scroll out to find the route unless you choose addresses in San Francisco. There is another PR that will handle the readjustment of the map)
4. Add stops
5. Verify that new routes are generated correctly

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/30609178/fe1e55af-eb9f-4f88-9c73-f8256bd9d055

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/30609178/8d8a2f79-31ec-4f75-9e93-462f95f73d51

</details>

<details>
<summary>Mobile Web - Safari</summary>

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Comment by hayata-suenaga** [comment]
hey @thienlnam I sent you a message about this PR [in slack](https://expensify.slack.com/archives/C05DWUDHVK7/p1692392715514449?thread_ts=1692373463.311519&cid=C05DWUDHVK7)

**Comment by hayata-suenaga** [comment]
just heads up. I'm changing ` react-native-x-maps` a little bit in [this PR](https://github.com/Expensify/react-native-x-maps/pull/23) to fix some issues, but this doesn't change the interface/API of `MapView`. So you don't have to wait for it to finish to merge your PR 👍

**Comment by thienlnam** [comment]
Since @allroundexperts reviewed a similar PR here, I asked if they were able to take it - though if you are around and can get to it earlier @eVoloshchak go for it as this is a high priority PR https://expensify.slack.com/archives/C02NK2DQWUX/p1692665981227269

File: src/components/DistanceRequest.js
Comment (thienlnam) [reply src/components/DistanceRequest.js:151]: There are multiple addresses, so I don't think this would work - we kind of need isEqual to check if all the waypoints are the same or not

**Comment by thienlnam** [comment]
> I'm always getting Please select a valid waypoint. error despite the addresses being correct. Is there anything that I'm missing?

Ah, we fixed something on the back-end. I think this might be related to that, what is the error you are getting on the back-end?

You might have to try on the staging server now

**Comment by thienlnam** [comment]
Added the usePrevious hook, ready to go

**Comment by thienlnam** [comment]
Yeah that is expected - it only fetches a route when you have valid waypoints (either have to remove the empty waypoints or add an address)

**Comment by thienlnam** [comment]
I would say so since you have the start and end waypoints. Let me check what google maps does

**Comment by thienlnam** [comment]
Ah I guess for google maps it doesn't show a route until everything is defined. I can do the same here

**Comment by thienlnam** [comment]
Okay, updated

**Comment by thienlnam** [comment]
Updated @allroundexperts

**Review by allroundexperts — APPROVED** [review]
Looks good!

I wasn't able to test this on Android Chrome because of an issue I mentioned [here](https://github.com/Expensify/App/pull/25161#issuecomment-1687122634).

File: src/components/DistanceRequest.js
Comment (hayata-suenaga) [inline src/components/DistanceRequest.js:95]: is there a specific reason we're not use `useEffect` with `waypoints` as a dependency?

**Review by hayata-suenaga — DISMISSED** [review]

File: src/components/DistanceRequest.js
Comment (thienlnam) [reply src/components/DistanceRequest.js:95]: https://github.com/Expensify/App/pull/25434#discussion_r1302152218

Updated it here - mostly so the logic within the useEffect is cleaner

Comment (hayata-suenaga) [reply src/components/DistanceRequest.js:95]: ah I meant just doing

```ts
useEffect(() => {
   // fetch route if necessary
   }, [waypoints]
  )
```

Comment (thienlnam) [reply src/components/DistanceRequest.js:95]: Hmm I'm not quite following,
GetRoute takes a transactionID so that is required with our linter to add in the useEffect

Do you mean only updating shouldFetchRoute with a useEffect?

Comment (hayata-suenaga) [reply src/components/DistanceRequest.js:95]: by the way, please merge this PR when you're ready. this is not a blocker 😄

Comment (hayata-suenaga) [reply src/components/DistanceRequest.js:95]: wait, I'm confused on another point. isn't the condition for `shouldFetchRoute` the opposite of the current one?

for example, `shouldFetchRoute` should be `true` when `isOffline` is true?

Comment (thienlnam) [reply src/components/DistanceRequest.js]: Ah you're right - I messed up the logic when I updated this

Updated the conditions

File: src/components/DistanceRequest.js
Comment (thienlnam) [reply src/components/DistanceRequest.js]: I fixed it - we don't use the inverse as much so I just adjusted the check

**Review by allroundexperts — APPROVED** [review]

**Comment by thienlnam** [comment]
Removing additional auto-assigned reviewers since @hayata-suenaga is already on this

**Review by hayata-suenaga — APPROVED** [review]

**Pull request merged**
